File: docs/js/services/learningProgressManager.js
/**
 * Learning Progress Manager - Simple and clean implementation
 *
 * Rules:
 * - Every 5 minutes → goal increases by 1 word (12 words/hour)
 * - Complete 1 mastered word (7/10) → wordsLearned++
 * - Sync immediately to Appwrite
 * - Load from Appwrite on refresh
 */
class LearningProgressManager {
    constructor() {
        // Constants
        this.MINUTES_PER_WORD = 5; // 5 minutes = 1 word goal
        this.STORAGE_KEY = 'learning_progress';

        // State
        this.startTime = null;      // Session start time (milliseconds)
        this.wordsLearned = 0;      // Total words completed
        this.lastSyncTime = 0;      // Last sync timestamp
    }

    /**
     * Initialize - Load from Appwrite or create new session
     */
    async initialize() {
        console.log('🔄 [Progress] Initializing...');

        try {
            // Try to load from Appwrite
            await appwriteService.loginAnonymously();
            const serverData = await appwriteService.getLearningProgress();

            if (serverData && serverData.startTime) {
                // Load existing session from server
                this.startTime = parseInt(serverData.startTime);
                this.wordsLearned = parseInt(serverData.wordsLearned) || 0;
                console.log('✅ [Progress] Loaded from Appwrite:', {
                    startTime: new Date(this.startTime).toLocaleString(),
                    wordsLearned: this.wordsLearned,
                    goal: this.getCurrentGoal(),
                    debt: this.getDebt()
                });
            } else {
                // Create new session
                this.startTime = Date.now();
                this.wordsLearned = 0;
                console.log('🆕 [Progress] Created new session');

                // Sync new session to server
                await this.syncToAppwrite();
            }

            // Save to localStorage as backup
            this.saveToLocalStorage();

        } catch (error) {
            console.error('❌ [Progress] Init failed:', error);

            // Fallback to localStorage
            const localData = localStorage.getItem(this.STORAGE_KEY);
            if (localData) {
                const data = JSON.parse(localData);
                this.startTime = data.startTime || Date.now();
                this.wordsLearned = data.wordsLearned || 0;
                console.log('⚠️ [Progress] Using localStorage fallback');
            } else {
                // Last resort: create new session
                this.startTime = Date.now();
                this.wordsLearned = 0;
                console.log('🆕 [Progress] Created new local session');
            }
        }

        console.log('✅ [Progress] Initialized:', this.getSummary());
    }

    /**
     * Add completed word - Call this when user completes a sentence correctly
     */
    async addCompletedWord() {
        const oldCount = this.wordsLearned;
        this.wordsLearned++;
        const newCount = this.wordsLearned;

        console.log('✅ [Progress] Word completed:', oldCount, '→', newCount);
        console.log('   🎯 Goal:', this.getCurrentGoal());
        console.log('   💳 Debt:', this.getDebt());

        // Save locally
        this.saveToLocalStorage();

        // Sync to server immediately
        try {
            await this.syncToAppwrite();
            console.log('   ✅ Synced to Appwrite');
        } catch (error) {
            console.error('   ❌ Sync failed:', error.message);
        }

        return newCount;
    }

    /**
     * Calculate current goal based on elapsed time
     * Formula: elapsedMinutes / 5 = goal
     */
    getCurrentGoal() {
        const elapsedMinutes = this.getElapsedMinutes();
        const goal = Math.floor(elapsedMinutes / this.MINUTES_PER_WORD);
        return goal;
    }

    /**
     * Get elapsed time in minutes
     */
    getElapsedMinutes() {
        if (!this.startTime) return 0;
        const elapsed = Date.now() - this.startTime;
        return elapsed / (60 * 1000);
    }

    /**
     * Get elapsed time formatted as HH:MM:SS
     */
    getElapsedTimeFormatted() {
        if (!this.startTime) return '0:00';

        const elapsed = Date.now() - this.startTime;
        const hours = Math.floor(elapsed / (60 * 60 * 1000));
        const minutes = Math.floor((elapsed % (60 * 60 * 1000)) / (60 * 1000));
        const seconds = Math.floor((elapsed % (60 * 1000)) / 1000);

        if (hours > 0) {
            return `${hours}:${minutes.toString().padStart(2, '0')}:${seconds.toString().padStart(2, '0')}`;
        }
        return `${minutes}:${seconds.toString().padStart(2, '0')}`;
    }

    /**
     * Get debt (how many words behind schedule)
     */
    getDebt() {
        const goal = this.getCurrentGoal();
        const debt = Math.max(0, goal - this.wordsLearned);
        return debt;
    }

    /**
     * Get progress percentage
     */
    getProgressPercentage() {
        const goal = this.getCurrentGoal();
        if (goal === 0) return 0;
        const percentage = Math.min(100, Math.floor((this.wordsLearned / goal) * 100));
        return percentage;
    }

    /**
     * Get level based on progress
     */
    getLevel() {
        const percentage = this.getProgressPercentage();

        if (percentage >= 100) return 'Master';
        if (percentage >= 80) return 'Advanced';
        if (percentage >= 60) return 'Intermediate';
        if (percentage >= 40) return 'Beginner';
        if (percentage >= 20) return 'Novice';
        return 'Starting';
    }

    /**
     * Get time status class (for UI coloring)
     */
    getTimeStatusClass() {
        const debt = this.getDebt();
        if (debt === 0) return 'success';
        if (debt < 5) return 'warning';
        return 'danger';
    }

    /**
     * Get summary object (for UI display)
     */
    getSummary() {
        return {
            wordsLearned: this.wordsLearned,
            goal: this.getCurrentGoal(),
            debt: this.getDebt(),
            progressPercentage: this.getProgressPercentage(),
            level: this.getLevel(),
            elapsedTime: this.getElapsedTimeFormatted(),
            timeStatusClass: this.getTimeStatusClass()
        };
    }

    /**
     * Save to localStorage (backup)
     */
    saveToLocalStorage() {
        const data = {
            startTime: this.startTime,
            wordsLearned: this.wordsLearned,
            lastUpdated: Date.now()
        };
        localStorage.setItem(this.STORAGE_KEY, JSON.stringify(data));
    }

    /**
     * Sync to Appwrite
     */
    async syncToAppwrite() {
        if (!this.startTime) {
            console.warn('⚠️ [Progress] Cannot sync: startTime is null');
            return;
        }

        await appwriteService.saveLearningProgress({
            startTime: this.startTime,
            wordsLearned: this.wordsLearned
        });

        this.lastSyncTime = Date.now();
    }

    /**
     * Reset progress (start new session)
     */
    async resetProgress() {
        console.log('🔄 [Progress] Resetting...');

        this.startTime = Date.now();
        this.wordsLearned = 0;

        this.saveToLocalStorage();
        await this.syncToAppwrite();

        console.log('✅ [Progress] Reset complete');
    }

    /**
     * Debug info
     */
    debug() {
        const summary = this.getSummary();
        console.log('🐛 [Progress] Debug Info:');
        console.table(summary);
        console.log('   Start time:', new Date(this.startTime).toLocaleString());
        console.log('   Elapsed minutes:', this.getElapsedMinutes().toFixed(2));
        return summary;
    }
}

// Global instance
const learningProgressManager = new LearningProgressManager();

// Debug command
window.debugProgress = () => learningProgressManager.debug();
window.resetProgress = () => learningProgressManager.resetProgress();

File: app/src/main/java/com/example/specialenglishlearningapp/utils/LearningProgressManager.kt
package com.example.specialenglishlearningapp.utils

import android.content.Context
import com.example.specialenglishlearningapp.constants.AppwriteConfig
import io.appwrite.ID
import io.appwrite.Query
import io.appwrite.services.Databases
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

object LearningProgressManager {
    private const val MINUTES_PER_WORD = 5
    private const val WORDS_PER_HOUR = 60 / MINUTES_PER_WORD
    private const val PROGRESS_DOCUMENT_ID = "user_learning_progress" // Fixed ID for single-user progress

    private var cachedStartTime: Long? = null
    private var cachedWordsLearned: Int? = null
    private var lastSyncTime: Long = 0
    private const val SYNC_INTERVAL_MS = 30000L // Sync every 30 seconds

    /**
     * Initialize learning progress from Appwrite
     * Call this when app starts
     */
    suspend fun initialize(context: Context): Result<Unit> = withContext(Dispatchers.IO) {
        try {
            Logger.d("🔄 [LearningProgress] Initializing...")
            val appwriteHelper = AppwriteHelper.getInstance(context)

            // Ensure authentication
            Logger.d("🔐 [LearningProgress] Checking authentication...")
            val currentUser = appwriteHelper.getCurrentUser()
            if (currentUser == null) {
                Logger.d("🔐 [LearningProgress] No session, logging in anonymously...")
                val session = appwriteHelper.loginAnonymously()
                Logger.d("✅ [LearningProgress] Login successful: userId=${session.userId}")
            } else {
                Logger.d("✅ [LearningProgress] Already authenticated: userId=${currentUser.id}")
            }

            val databases = appwriteHelper.databases

            // Try to fetch existing progress
            try {
                Logger.d("📥 [LearningProgress] Fetching from Appwrite...")
                Logger.d("   Database: ${AppwriteConfig.DATABASE_ID}")
                Logger.d("   Collection: ${AppwriteConfig.LEARNING_PROGRESS_COLLECTION_ID}")
                Logger.d("   Document: $PROGRESS_DOCUMENT_ID")

                val document = databases.getDocument(
                    databaseId = AppwriteConfig.DATABASE_ID,
                    collectionId = AppwriteConfig.LEARNING_PROGRESS_COLLECTION_ID,
                    documentId = PROGRESS_DOCUMENT_ID
                )

                val data = document.data
                cachedStartTime = (data["startTime"] as? String)?.toLongOrNull() ?: System.currentTimeMillis()
                cachedWordsLearned = (data["wordsLearned"] as? String)?.toIntOrNull() ?: 0
                lastSyncTime = System.currentTimeMillis()

                Logger.d("✅ [LearningProgress] Loaded from Appwrite successfully!")
                Logger.d("   📅 startTime: $cachedStartTime")
                Logger.d("   📚 wordsLearned: $cachedWordsLearned")
                Logger.d("   ⏱️ elapsed: ${formatElapsedTime()}")
                Logger.d("   🎯 currentGoal: ${getCurrentGoal()}")
                Logger.d("   💳 debt: ${getDebt()}")
            } catch (e: Exception) {
                // Document doesn't exist, create new one
                Logger.d("⚠️ [LearningProgress] Document not found: ${e.message}")
                Logger.d("📝 [LearningProgress] Creating new document...")

                val startTime = System.currentTimeMillis()
                val data = mapOf(
                    "startTime" to startTime.toString(),
                    "wordsLearned" to "0",
                    "lastUpdated" to startTime.toString()
                )

                databases.createDocument(
                    databaseId = AppwriteConfig.DATABASE_ID,
                    collectionId = AppwriteConfig.LEARNING_PROGRESS_COLLECTION_ID,
                    documentId = PROGRESS_DOCUMENT_ID,
                    data = data
                )

                cachedStartTime = startTime
                cachedWordsLearned = 0
                lastSyncTime = System.currentTimeMillis()

                Logger.d("✅ [LearningProgress] Document created successfully!")
                Logger.d("   📅 startTime: $startTime")
                Logger.d("   📚 wordsLearned: 0")
            }

            Logger.d("🎉 [LearningProgress] Initialization complete!")
            Result.success(Unit)
        } catch (e: Exception) {
            Logger.e("❌ [LearningProgress] Initialization failed: ${e.message}", e)
            e.printStackTrace()
            // Fallback to local time if Appwrite fails
            cachedStartTime = cachedStartTime ?: System.currentTimeMillis()
            cachedWordsLearned = cachedWordsLearned ?: 0
            Result.failure(e)
        }
    }

    /**
     * Get start time (from cache or Appwrite)
     *
     * ⚠️ QUAN TRỌNG: startTime KHÔNG BAO GIỜ được thay đổi sau khi khởi tạo.
     * Nó xác định điểm bắt đầu tính toán mục tiêu và PHẢI BẤT BIẾN.
     */
    fun getStartTime(): Long {
        return cachedStartTime ?: System.currentTimeMillis()
    }

    /**
     * Get elapsed time in milliseconds since start
     */
    fun getElapsedTimeMillis(): Long {
        return System.currentTimeMillis() - getStartTime()
    }

    /**
     * Get elapsed time in hours
     */
    fun getElapsedHours(): Float {
        return getElapsedTimeMillis() / (60 * 60 * 1000f)
    }

    /**
     * Get elapsed time in minutes
     */
    fun getElapsedMinutes(): Float {
        return getElapsedTimeMillis() / (60 * 1000f)
    }

    fun getCurrentGoal(): Int {
        val elapsedMinutes = getElapsedMinutes()
        val calculatedGoal = (elapsedMinutes / MINUTES_PER_WORD).toInt().coerceAtLeast(1)
        Logger.d("Current goal: $calculatedGoal words (elapsed: ${String.format("%.2f", elapsedMinutes)} minutes = ${String.format("%.2f", getElapsedHours())} hours)")
        return calculatedGoal
    }

    /**
     * Get total words learned (from cache)
     */
    fun getWordsLearned(): Int {
        return cachedWordsLearned ?: 0
    }

    /**
     * Calculate debt (words owed)
     * Nợ = Mục tiêu hiện tại - Số từ đã học
     */
    fun getDebt(): Int {
        val debt = getCurrentGoal() - getWordsLearned()
        return debt.coerceAtLeast(0) // Never negative
    }

    /**
     * Get progress percentage
     */
    fun getProgressPercentage(): Int {
        val goal = getCurrentGoal()
        val learned = getWordsLearned()
        return if (goal > 0) (learned * 100 / goal).coerceAtMost(100) else 0
    }

    /**
     * Format elapsed time as HH:MM:SS
     */
    fun formatElapsedTime(): String {
        val elapsed = getElapsedTimeMillis()
        val hours = elapsed / (60 * 60 * 1000)
        val minutes = (elapsed % (60 * 60 * 1000)) / (60 * 1000)
        val seconds = (elapsed % (60 * 1000)) / 1000
        return String.format("%02d:%02d:%02d", hours, minutes, seconds)
    }

    /**
     * Add completed vocabulary and sync to Appwrite
     * Auto-sync every 30 seconds to ensure data is saved
     *
     * ✅ ĐẢMBẢO: Hàm này CHỈ TĂNG wordsLearned, KHÔNG BAO GIỜ GIẢM.
     */
    suspend fun addCompletedVocabulary(context: Context, forceSync: Boolean = false): Result<Unit> = withContext(Dispatchers.IO) {
        try {
            cachedWordsLearned = (cachedWordsLearned ?: 0) + 1  // CHỈ TĂNG (+1)
            Logger.d("📚 Word completed. Total learned: $cachedWordsLearned")
            Logger.d("   Current goal: ${getCurrentGoal()}")
            Logger.d("   Debt: ${getDebt()}")

            // Always try to sync after completing a word
            val shouldSync = forceSync || (System.currentTimeMillis() - lastSyncTime) >= SYNC_INTERVAL_MS

            if (shouldSync) {
                Logger.d("🔄 Auto-syncing to Appwrite...")
                val syncResult = syncToAppwrite(context)
                if (syncResult.isFailure) {
                    Logger.e("⚠️ Sync failed but progress saved in memory: ${syncResult.exceptionOrNull()?.message}")
                }
            } else {
                val remainingSeconds = (SYNC_INTERVAL_MS - (System.currentTimeMillis() - lastSyncTime)) / 1000
                Logger.d("⏳ Sync scheduled in ${remainingSeconds}s")
            }

            Result.success(Unit)
        } catch (e: Exception) {
            Logger.e("Failed to add completed vocabulary", e)
            Result.failure(e)
        }
    }

    /**
     * Force sync to Appwrite
     */
    suspend fun syncToAppwrite(context: Context): Result<Unit> = withContext(Dispatchers.IO) {
        try {
            Logger.d("☁️ [LearningProgress] Starting sync to Appwrite...")
            val appwriteHelper = AppwriteHelper.getInstance(context)

            // Ensure authentication
            val currentUser = appwriteHelper.getCurrentUser()
            if (currentUser == null) {
                Logger.d("🔐 [LearningProgress] Re-authenticating...")
                val session = appwriteHelper.loginAnonymously()
                Logger.d("✅ [LearningProgress] Login successful: userId=${session.userId}")
            } else {
                Logger.d("✅ [LearningProgress] Already authenticated: userId=${currentUser.id}")
            }

            val databases = appwriteHelper.databases

            val data = mapOf(
                "startTime" to getStartTime().toString(),
                "wordsLearned" to getWordsLearned().toString(),
                "lastUpdated" to System.currentTimeMillis().toString()
            )

            Logger.d("📤 [LearningProgress] Sending data:")
            Logger.d("   Database: ${AppwriteConfig.DATABASE_ID}")
            Logger.d("   Collection: ${AppwriteConfig.LEARNING_PROGRESS_COLLECTION_ID}")
            Logger.d("   Document: $PROGRESS_DOCUMENT_ID")
            Logger.d("   startTime: ${getStartTime()}")
            Logger.d("   wordsLearned: ${getWordsLearned()}")
            Logger.d("   currentGoal: ${getCurrentGoal()}")
            Logger.d("   debt: ${getDebt()}")

            try {
                // Try to update existing document
                Logger.d("🔄 [LearningProgress] Attempting to update document...")
                val result = databases.updateDocument(
                    databaseId = AppwriteConfig.DATABASE_ID,
                    collectionId = AppwriteConfig.LEARNING_PROGRESS_COLLECTION_ID,
                    documentId = PROGRESS_DOCUMENT_ID,
                    data = data
                )
                Logger.d("✅ [LearningProgress] Updated document successfully!")
                Logger.d("   Response ID: ${result.id}")
            } catch (e: Exception) {
                // Document doesn't exist, create it
                Logger.d("⚠️ [LearningProgress] Update failed: ${e.message}")
                Logger.d("📝 [LearningProgress] Creating new document...")
                val result = databases.createDocument(
                    databaseId = AppwriteConfig.DATABASE_ID,
                    collectionId = AppwriteConfig.LEARNING_PROGRESS_COLLECTION_ID,
                    documentId = PROGRESS_DOCUMENT_ID,
                    data = data
                )
                Logger.d("✅ [LearningProgress] Created document successfully!")
                Logger.d("   Response ID: ${result.id}")
            }

            lastSyncTime = System.currentTimeMillis()
            Logger.d("🎉 [LearningProgress] Sync completed at: $lastSyncTime")
            Result.success(Unit)
        } catch (e: Exception) {
            Logger.e("❌ [LearningProgress] Sync failed: ${e.message}", e)
            Logger.e("   Exception type: ${e.javaClass.name}")
            Logger.e("   Stack trace: ${e.stackTraceToString()}")
            e.printStackTrace()
            Result.failure(e)
        }
    }

    /**
     * Reset progress (DISABLED - Mục tiêu không bao giờ được reset)
     *
     * HÀM NÀY ĐÃ BỊ VÔ HIỆU HÓA để đảm bảo mục tiêu luôn tăng dần
     * và không bao giờ bị reset về 0.
     */
    @Deprecated("Hàm này đã bị vô hiệu hóa. Mục tiêu không được phép reset.", level = DeprecationLevel.ERROR)
    suspend fun resetProgress(context: Context): Result<Unit> = withContext(Dispatchers.IO) {
        Logger.e("⛔ resetProgress() is DISABLED - Goals must never be reset!")
        Result.failure(UnsupportedOperationException("Reset progress is disabled. Goals must always increase."))
    }

    /**
     * Get level based on words learned
     */
    fun getLevel(): String {
        val learned = getWordsLearned()
        return when {
            learned >= 1000 -> "🏆 Bậc thầy"
            learned >= 500 -> "🌟 Chuyên gia"
            learned >= 200 -> "⭐ Trung cấp"
            learned >= 50 -> "📚 Đang tiến bộ"
            else -> "🌱 Mới bắt đầu"
        }
    }

    /**
     * Get detailed progress info
     */
    fun getProgressDetails(): String {
        val learned = getWordsLearned()
        val goal = getCurrentGoal()
        val debt = getDebt()
        val percentage = getProgressPercentage()
        val elapsed = formatElapsedTime()

        return buildString {
            append("Đã học: $learned/$goal từ (${percentage}%)\n")
            if (debt > 0) {
                append("Còn thiếu: $debt từ\n")
            } else {
                append("Đã hoàn thành mục tiêu!\n")
            }
            append("Thời gian: $elapsed")
        }
    }

    /**
     * Test Appwrite connection and permissions
     */
    suspend fun testAppwriteConnection(context: Context): Result<String> = withContext(Dispatchers.IO) {
        try {
            Logger.d("🧪 [Test] Testing Appwrite connection...")
            val appwriteHelper = AppwriteHelper.getInstance(context)

            // Test 1: Check authentication
            Logger.d("🧪 [Test] Step 1: Checking authentication...")
            val currentUser = appwriteHelper.getCurrentUser()
            if (currentUser == null) {
                Logger.d("🧪 [Test] No user, logging in anonymously...")
                val session = appwriteHelper.loginAnonymously()
                Logger.d("✅ [Test] Logged in: userId=${session.userId}")
            } else {
                Logger.d("✅ [Test] Already logged in: userId=${currentUser.id}")
            }

            // Test 2: Try to read document
            Logger.d("🧪 [Test] Step 2: Reading document...")
            val databases = appwriteHelper.databases
            try {
                val doc = databases.getDocument(
                    databaseId = AppwriteConfig.DATABASE_ID,
                    collectionId = AppwriteConfig.LEARNING_PROGRESS_COLLECTION_ID,
                    documentId = PROGRESS_DOCUMENT_ID
                )
                Logger.d("✅ [Test] Document found: ${doc.id}")
            } catch (e: Exception) {
                Logger.d("⚠️ [Test] Document not found (will create): ${e.message}")
            }

            // Test 3: Try to write/update
            Logger.d("🧪 [Test] Step 3: Writing test data...")
            val testData = mapOf(
                "startTime" to System.currentTimeMillis().toString(),
                "wordsLearned" to "999",
                "lastUpdated" to System.currentTimeMillis().toString()
            )

            try {
                val result = databases.updateDocument(
                    databaseId = AppwriteConfig.DATABASE_ID,
                    collectionId = AppwriteConfig.LEARNING_PROGRESS_COLLECTION_ID,
                    documentId = PROGRESS_DOCUMENT_ID,
                    data = testData
                )
                Logger.d("✅ [Test] Update successful: ${result.id}")
                Result.success("✅ Test thành công!\nDocument ID: ${result.id}")
            } catch (e: Exception) {
                Logger.d("⚠️ [Test] Update failed, trying create: ${e.message}")
                val result = databases.createDocument(
                    databaseId = AppwriteConfig.DATABASE_ID,
                    collectionId = AppwriteConfig.LEARNING_PROGRESS_COLLECTION_ID,
                    documentId = PROGRESS_DOCUMENT_ID,
                    data = testData
                )
                Logger.d("✅ [Test] Create successful: ${result.id}")
                Result.success("✅ Test thành công (tạo mới)!\nDocument ID: ${result.id}")
            }
        } catch (e: Exception) {
            Logger.e("❌ [Test] Failed: ${e.message}", e)
            Result.failure(e)
        }
    }
}
